src/compiler/checker.ts

```diff
@@ -18323,7 +18323,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 // then no instantiations will be and we can just return the false branch type.
                 if (!(inferredExtendsType.flags & TypeFlags.AnyOrUnknown) && (checkType.flags & TypeFlags.Any || !isTypeAssignableTo(getPermissiveInstantiation(checkType), getPermissiveInstantiation(inferredExtendsType)))) {
                     // Return union of trueType and falseType for 'any' since it matches anything
-                    if (checkType.flags & TypeFlags.Any) {
+                    if (checkType.flags & TypeFlags.Any || forConstraint && isTypeAssignableTo(getPermissiveInstantiation(inferredExtendsType), getPermissiveInstantiation(checkType))) {

```

Comment:
While this is probably a bit better than the often-incorrect result we give now, so maybe we're just fine with that being that, isn't a check in this style kind of fundamentally flawed? For example, a `type Cond<T extends {a: string} | {b: string}> = T extends {c: string} ? true : false`. `T`'s constraint isn't related in any way to the check type of `{c: string}`, yet the constraint should probably still be the union of both branches, as an instantiation of `T` might satisfy the check type, or might not, while here we're still skipping the `true` branch type just because `{a: string}` isn't assignable to `{c: string}`. And comparability wouldn't have been any better. As @andrewbranch said - the only case where we can definitively say only one branch is taken would seem to require checking for an empty intersection between the check-type-constraint and the extends type (and, as you mentioned, ideally tighter empty intersection logic so that has higher accuracy).

How different are the results it if we just replace this assignability check with a `intersectTypes(getPermissiveInstantiation(inferredExtendsType), getPermissiveInstantiation(checkType)) !== neverType`? Are the bad changes any worse?